server/src/main/scala/com/cloudera/livy/server/LivyServer.scala

```diff
@@ -126,6 +142,29 @@ class LivyServer extends Logging {
     sys.props("livy.server.serverUrl") = _serverUrl.get
   }
 
+  def runKinit(livyConf: LivyConf): Runnable = new Runnable {
+    override def run(): Unit = {
+      UserGroupInformation.getCurrentUser.reloginFromTicketCache()
+      val keytab = livyConf.get(LAUNCH_KERBEROS_KEYTAB)
+      val principal = livyConf.get(LAUNCH_KERBEROS_PRINCIPAL)
+      require(principal != null,
+        s"Kerberos requires ${LAUNCH_KERBEROS_KEYTAB.key} to be provided.")
+      require(keytab != null,
+        s"Kerberos requires ${LAUNCH_KERBEROS_PRINCIPAL.key} to be provided.")
+      val commands = ArrayBuffer.empty[String]
+      commands += "kinit"
+      commands += "-kt"
+      commands += keytab
+      commands += principal
+      val proc = new ProcessBuilder(commands: _*).start()
+      proc.waitFor() match {
+        case 0 => info("Run kinit command")
+        case _ => warn("Fail to run kinit command," +
```

Comment:
This looks like a very serious error that should be fatal; maybe allow a certain number of failures, but shut down the server after that. The first login should definitely be required (server shouldn't even start if kinit fails).